plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -212,17 +212,21 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
             -- actually have got rid of all of them!
             pure $ mkLet ann NonRec bs' t'
         -- This includes recursive let terms, we don't even consider inlining them at the moment
-        t ->
-            -- process all subterms first, so that the rhs won't be processed more than once. This
-            -- is important because otherwise the number of times we process them can grow
-            -- exponentially in the case that it has nested `let`s.
-            --
-            -- Then, consider call site inlining for each node that have gone through unconditional
-            -- inlining. Because `inlineSaturatedApp` traverses *all* application nodes for each
-            -- subterm, the runtime is quadratic for terms with a long chain of applications.
-            -- If we use the context-based approach like in GHC, this won't be a problem, so we may
-            -- consider that in the future.
-            inlineSaturatedApp =<< forMOf termSubterms t processTerm
+        t -> do
+            {- We consider call site inlining when we see an application node, then process the term
+            (perform unconditional inlining). Because `inlineApp` traverses *all* application nodes
+            for each subterm, the runtime is quadratic for terms with a long chain of applications.
+            If we use the context-based approach like in GHC, this won't be a problem, so we may
+            consider that in the future.
+
+            Ideally, we would process all subterms first, so that the rhs won't be processed more
+            than once. This is important because otherwise the number of times we process them can
+            grow exponentially in the case that it has nested `let`s.
+            However, if we do that then `CallSiteInline.inlineApp` would be given a subterm of each
+            application node, instead of a whole application chain, and it would not reduce
+            properly.-}
+            t' <- inlineApp t
+            forMOf termSubterms t' processTerm
```

Comment:
Note for reviewers, this was the fix for the last set of problems. The issue is that the old behaviour would process the subterms first, so would always look at the _prefix_ of an application before the full application. This often led to it not beta-reducing enough. The problem didn't arise before because we only ever inlined fully-saturated applications, so it didn't matter that we looked at the prefix first.

Reply:
Aha, I think I see a sensible way out:
- Split the term into an application context
- Process the sub-parts of the _application context_
- Now consider inlining the application context

Reply:
Trying. Although this probably wouldn't fix the regressions.

Reply:
But `(\y -> 1) 2` is not a problem, right? It will be reduced to 2 in the next iteration of the simplifier.
Yes, it requires one more iteration in this instance. But simplifying a term without simplifying its subterms can also lead to needing additional iterations.

Reply:
I don't know if it's a problem, but it very much seems to me to be not what the code is _trying_ to do. It's pretty counter-intuitive and I had to think about it a lot. Why would we _not_ reduce the longest application that we can, given we have all the machinery for reducing multiple applications right there!

Reply:
Ok, fair enough.